[fsdp2] simplify fsdp_param logic with DTensorSpec

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #128242

as titled, we can use a single DTensorSpec to save the SPMD sharding
spec, plus the global shape/stride to simplify the FSDPParam logic

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k

wanchaol committed:
[fsdp2] simplify fsdp_param logic with DTensorSpec

as titled, we can use a single DTensorSpec to save the SPMD sharding
spec, plus the global shape/stride to simplify the FSDPParam logic

[ghstack-poisoned]

awgu commented on torch/distributed/_composable/fsdp/_fsdp_param.py:
```diff
@@ -207,28 +203,46 @@ def _init_sharded_param(self, param: nn.Parameter, device: torch.device):
                 self._spmd_placements = (Shard(0),)
-            self._global_size = param.size()
-            self._global_stride = param.stride()
+            # self._global_size = param.size()
+            # self._global_stride = param.stride()
+            self._sharding_spec = DTensorSpec(
```
nit: remember to remove

wanchaol committed:
Update on "[fsdp2] simplify fsdp_param logic with DTensorSpec"


as titled, we can use a single DTensorSpec to save the SPMD sharding
spec, plus the global shape/stride to simplify the FSDPParam logic

cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k

[ghstack-poisoned]

wanchaol committed:
Update on "[fsdp2] simplify fsdp_param logic with DTensorSpec"


as titled, we can use a single DTensorSpec to save the SPMD sharding
spec, plus the global shape/stride to simplify the FSDPParam logic

cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k

[ghstack-poisoned]

wanchaol committed:
Update on "[fsdp2] simplify fsdp_param logic with DTensorSpec"


as titled, we can use a single DTensorSpec to save the SPMD sharding
spec, plus the global shape/stride to simplify the FSDPParam logic

cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k

[ghstack-poisoned]

wanchaol committed:
Update on "[fsdp2] simplify fsdp_param logic with DTensorSpec"


as titled, we can use a single DTensorSpec to save the SPMD sharding
spec, plus the global shape/stride to simplify the FSDPParam logic

cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k

[ghstack-poisoned]

wanchaol commented:
@pytorchbot merge

wanchaol commented:
@pytorchbot merge

wanchaol commented:
@pytorchbot merge -i

wanchaol commented:
@pytorchbot merge -f "failures not related"
